src/RA_Dlg_MemBookmark.h

```diff
@@ -8,93 +8,169 @@
+    {
+        return{ a.m_nAddress < b.m_nAddress };
+    }
+
+    friend auto swap(_Inout_ MemBookmark& a, _Inout_ MemBookmark& b) noexcept
+        ->decltype(std::swap(a.m_nAddress, b.m_nAddress));
 };
 
+// No return doesn't work everywhere, it must be a bug
+
 class Dlg_MemBookmark
 {
+    // one is const and one isn't const, isn't that an error in logic? Like if I
+    // put Bookmarks as the mapped type it won't be const anymore.
+    using MyBookmarks    = std::vector<MemBookmark>;
+    using MyBookmarkMap  = std::map<ByteAddress, MyBookmarks>;
+    using iterator       = MyBookmarkMap::iterator;
+    using const_iterator = MyBookmarkMap::const_iterator;
+
 public:
-    //Dlg_MemBookmark();
-    //~Dlg_MemBookmark();
+    // compiler generated constructor throws on all except move assignment and
+    // destructor so we can't use = default - Samer
+    Dlg_MemBookmark() noexcept {};
+    ~Dlg_MemBookmark() noexcept = default;
+    Dlg_MemBookmark(const Dlg_MemBookmark&) = delete;
+    // probably doesn't need moving either
+    Dlg_MemBookmark& operator=(const Dlg_MemBookmark&) = delete;
+    Dlg_MemBookmark(Dlg_MemBookmark&&) noexcept = delete;
+    Dlg_MemBookmark& operator=(Dlg_MemBookmark&&) noexcept = delete;
+
+
 
-    static INT_PTR CALLBACK s_MemBookmarkDialogProc(HWND, UINT, WPARAM, LPARAM);
-    INT_PTR MemBookmarkDialogProc(HWND, UINT, WPARAM, LPARAM);
+    _NODISCARD static INT_PTR CALLBACK s_MemBookmarkDialogProc(_In_ HWND, _In_ UINT,
+        _In_ WPARAM, _In_ LPARAM);
+    _NODISCARD INT_PTR CALLBACK MemBookmarkDialogProc(_In_ HWND, _In_ UINT,
+        _In_ WPARAM, _In_ LPARAM);
 
-    void InstallHWND(HWND hWnd) { m_hMemBookmarkDialog = hWnd; }
-    HWND GetHWND() const { return m_hMemBookmarkDialog; }
-    BOOL IsActive() const;
+    void InstallHWND(_In_ HWND hWnd) { m_hMemBookmarkDialog = hWnd; }
+    _NODISCARD _CONSTANT_FN& GetHWND() const { return m_hMemBookmarkDialog; }
+    _NODISCARD BOOL IsActive() const;
 
-    std::vector<MemBookmark*> Bookmarks() { return m_vBookmarks; }
-    void UpdateBookmarks(bool bForceWrite);
-    void AddBookmark(MemBookmark* newBookmark) { m_vBookmarks.push_back(newBookmark); }
-    void WriteFrozenValue(const MemBookmark& Bookmark);
+    // Needs non-const reference version for EditProcBM
+    _NODISCARD auto& Bookmarks() { return m_vBookmarks; }
+    _NODISCARD auto& Bookmarks() const { return m_vBookmarks; }
+
+    _NODISCARD auto end() const noexcept
+        ->decltype(std::declval<MyBookmarkMap&>().cend())
+    {
+        return m_BookmarkMap.cend();
+    }
+
+    void UpdateBookmarks(_In_ bool bForceWrite);
+    auto AddBookmark(MemBookmark&& newBookmark) noexcept // ok it's void
+        ->decltype(std::declval<MyBookmarks&>().push_back(std::move(newBookmark)))
+    {
+        MemBookmark myClone;
+        newBookmark.Clone(myClone);
+        m_vBookmarks.push_back(std::move(newBookmark));
+        AddBookmarkMap(std::move(myClone));
+
```

Comment:
The clone absolutely refused to enter AddBookmarkMap after AddBookmark even though it was valid. It does make sense though. If "Bookmarks" are a part of "BookmarkMap" then the Bookmarks need to be added after it's updated.